rclrs/src/node/client.rs

```diff
@@ -0,0 +1,257 @@
+                &mut sequence_number,
+            )
+        };
+        let requests = &mut *self.requests.lock();
+        requests.insert(sequence_number, Box::new(callback));
+        self.sequence_number.swap(sequence_number, Ordering::SeqCst);
+        ret.ok()
+    }
+
+    /// Send a request with a callback as a parameter.
+    ///
+    /// The [`MessageCow`] trait is implemented by any
+    /// [`Message`] as well as any reference to a `Message`.
+    ///
+    /// The reason for allowing owned messages is that publishing owned messages can be more
+    /// efficient in the case of idiomatic messages[^note].
+    ///
+    /// [^note]: See the [`Message`] trait for an explanation of "idiomatic".
+    ///
+    /// Hence, when a message will not be needed anymore after publishing, pass it by value.
+    /// When a message will be needed again after publishing, pass it by reference, instead of cloning and passing by value.
+    pub async fn call_async<'a, R: MessageCow<'a, T::Request>>(
+        &self,
+        request: R,
+    ) -> Result<T::Response, RclrsError>
+    where
+        T: rosidl_runtime_rs::Service + 'static,
+    {
+        let rmw_message = T::Request::into_rmw_message(request.into_cow());
+        let mut sequence_number = self.sequence_number.load(Ordering::SeqCst);
+        let ret = unsafe {
+            rcl_send_request(
+                &*self.handle.lock() as *const _,
+                rmw_message.as_ref() as *const <T::Request as Message>::RmwMsg as *mut _,
+                &mut sequence_number,
+            )
+        };
+        debug_assert_eq!(ret, 0);
+        let (tx, rx) = oneshot::channel::<T::Response>();
+        self.futures.lock().insert(sequence_number, tx);
+        Ok(rx.await?)
+    }
+
+    /// Ask RMW for the data
+    ///
+    /// +----------------------+
+    /// | rclrs::take_response |
+    /// +----------+-----------+
+    ///            |
+    ///            |
+    /// +----------v-----------+
+    /// |   rcl_take_response  |
+    /// +----------+-----------+
+    ///            |
+    ///            |
+    /// +----------v----------+
+    /// |      rmw_take       |
+    /// +---------------------+
+    pub fn take_response(&self) -> Result<(T::Response, rmw_request_id_t), RclrsError> {
+        let mut request_id_out = rmw_request_id_t {
+            writer_guid: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+            sequence_number: 0,
+        };
+        type RmwMsg<T> =
+            <<T as rosidl_runtime_rs::Service>::Response as rosidl_runtime_rs::Message>::RmwMsg;
+        let mut response_out = RmwMsg::<T>::default();
+        let handle = &mut *self.handle.lock();
+        let ret = unsafe {
+            rcl_take_response(
+                handle as *const _,
+                &mut request_id_out,
+                &mut response_out as *mut RmwMsg<T> as *mut _,
+            )
+        };
+        ret.ok()?;
```

Comment:
In the existing code we have the `ok()?` tacked on right after the function call. I like it because it means there is no danger of forgetting, or accidentally inserting a new line before `ret.ok()?` that continues to do stuff even though an error occurred.

Reply:
Not always, for example publisher and subscription use `let ret = unsafe { ... }; ret.ok?'`, so that code should be updated.

Reply:
Yes, it should.

Reply:
Wanna open a PR for that or shall I do it myself?